### DISCUSSION TITLE
The code of code blocks copied from PDFs is not correct

### DISCUSSION THREAD

**Discussion by fuhrmanator** [quarto-dev/quarto-cli#4008] [Q&A] [answered]
I use code blocks a lot for my course material, and so students want to copy it to use. On the HTML versions of my Quarto content, it's no problem (the copy icon is awesome).

However, in PDF it's another story. Here's a simple example:

````markdown
# Some sample java code

```java
ArrayList a = new
    ArrayList();
```
`````

If I copy the code from a PDF of the above, and paste it into a text editor, I get:

```
ArrayLista=newArrayList();1
```

The missing line break will cause an error because two tokens run together. Plus, there's a `1` at the end which I have no idea where it comes from.

It's probably (surely?) related to LaTeX, and there are several discussions I found, but didn't have time to try out all the options:
- https://tex.stackexchange.com/q/30284/17868
- https://tex.stackexchange.com/q/563803/17868

Any advice (using a different way to display code in LaTeX that has more reliable copy/paste?) would be helpful.

**Comment by dragonstyle** [comment, marked as answer]
One quick thing for you to try - I saw the same paste behavior that you described using the preview that we generate, but when I open the PDF using the system PDF viewer, the code pasted correctly as:

```
ArrayList a = new
      ArrayList();
```

Do you see the same thing? Not sure whether students will be using the preview instance vs. a distributed PDF in a desktop viewer, but could potentially reduce the problem.

**Reply by fuhrmanator** [reply]
Indeed, I was testing my content using the Quarto previewer. 
Using Acrobat Reader (or Sumatra PDF), I get a better result, in that there are line breaks.
However, the indentation isn't working (but it's acceptable in my case).

**Reply by dragonstyle** [reply]
Glad to hear it isn't totally broken - in Preview on Mac is appears to preserve whitespace fwiw. I am perplexed at what the `1` is in the copied code in the previewer.

**Comment by kputschko** [comment]
Have you had any further luck in solving this issues? I'm also distributing course material via PDF to students and would prefer to have them copy and paste easily from the document, regardless of the application they are using to view the documents.

**Reply by GitHubRulesOK** [reply]
@kputschko 
Commented on SO however PDF does not retain indents paragraphs line feeds tabs or other printer control codes (like my writing) as it is agnostic to which printer to be used on and screens dont use mechanical hints for text positioning.

PDF is one character at a time at a charted XY position thus the best extraction is one that ADDS line feeds on the right of a string of letters at same elevation and adds spaces to fill any voids on the left thus `POPPLER  PDFTOTEXT -LAYOUT` is best.
Second Best is Adobe export text for accessibility but sometimes there is no correlation with what is seen on the screen!

![image](https://github.com/user-attachments/assets/9923b20f-e9b1-4577-8359-d421ec9b9451)


The best a Reader could be expected to provide is left set text IF it is generated with spacebar characters.

![image](https://github.com/user-attachments/assets/9419970c-4021-49f3-be0c-b2f4e2ecd73d)


To get a reasonable output use the 3 decade Methods such as glyph and cog or ghostscript here is poppler PDFtoText

![image](https://github.com/user-attachments/assets/02161f81-0853-473c-846b-323d42d984ef)
